kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/KrpcFilterContext.java

```diff
@@ -38,12 +38,32 @@ public interface KrpcFilterContext {
     String sniHostname();
 
     /**
-     * Send a request towards the broker, invoking upstream filters.
+     * Creates a builder for a request filter result objects.  This object encapsulates
+     * the request to forward and optionally orders for actions such as closing
+     * the connection or dropping the request.
+     * <br/>
+     * The builder returns either {@link CompletionStage<RequestFilterResult>} object
+     * ready to be returned by the request filter, or a {@link ResponseFilterResult} object.
+     * The latter facilitates asynchronous programming patterns where requests must be
+     * forwarded after other work has completed.
+     *
+     * @return builder
+     */
+    RequestFilterResultBuilder requestFilterResultBuilder();
```

Comment:
I have been wondering if there is some benefit to obtaining the Result/CompletionStage from the KrpcFilterContext like this. Alternatively we could have the concrete Result/Builder classes live in the `api` module and the user could use them without reference to the Context. Like `return RequestFilterResultBuilder.builder().forward(header, body).build().completed()`. As it is the user could implement the *FilterResult interfaces themselves and supply unexpected combinations of object. Maybe if we had the concrete classes in the `api` we could prevent this mischief.

(this is something we could do as a future PR, it's another big refactor on top of what we've got)

Reply:
I guess a benefit is the context is visible immediately, rather than the user having to look around for the right way to build a result.

Reply:
I'm thinking validations like the following actually belong in the builder.

```
if (response.apiKey() != decodedFrame.apiKey().id) 
```

If that's correct then the builder's ctor of what it is actually building for.  That's an argument for keeping the context the factory for the builders.

I don't propose to add this change now.   I'd like to see us build this out test first.